File: examples/python/send_proof.py
import requests
from aligned_py.sdk import *
from aligned_py.core.types import *
import asyncio
from web3 import Web3

API_URL = "http://localhost:3030"
rpc_api = "https://ethereum-holesky-rpc.publicnode.com"
private_key = '0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80'

def open_file(elf_file_path):
    with open(elf_file_path, 'rb') as f:
        elf_data = f.read()
    return elf_data

def generate_proof(elf_file_path, input_data):
    url = f"{API_URL}/sp1/generate-proof-and-verify"

    elf_data = open_file(elf_file_path)
    input_bytes = input_data.encode() if isinstance(input_data, str) else input_data

    data = { "elf": list(elf_data), "input": list(input_bytes) }

    response = requests.post(url, json=data)
    response_json = response.json()

    if response_json.get("message") == "Incorrect prove":
        raise ValueError("Incorrect prove!")
    elif response_json.get("message") == "Output not added":
        raise ValueError("Output not added!")

    return response_json

async def main():
    web3 = Web3(Web3.HTTPProvider(rpc_api))
    account = web3.eth.account.from_key(private_key)
    nonce = get_next_nonce(rpc_api, account.address, Network.Holesky)
    max_fee = estimate_fee(rpc_api, PriceEstimate.Instant)
    print(f"Nonce: {nonce}")

    proof = generate_proof("examples/elf/riscv32im-succinct-zkvm-elf", "cca").get("proof")
    program_code = open_file("examples/elf/riscv32im-succinct-zkvm-elf")

    verification_data = VerificationData(
        proving_system=ProvingSystemId.SP1,
        proof=proof,
        public_input=None,
        verification_key=None,
        vm_program_code=bytearray(program_code),
        proof_generator_address=account.address
    )

    result = await submit_and_wait_verification(
        "wss://batcher.alignedlayer.com",
        rpc_api,
        Network.Holesky,
        verification_data,
        max_fee,
        account,
        nonce
    )
    print(f"0x{bytes(result.batch_merkle_root).hex()}")

asyncio.run(main())

File: examples/javascript/sendProof.js
import path from "path";
import { fileURLToPath } from "url";
import axios from 'axios';
import { ethers } from 'ethers';
import fs from 'fs/promises';
import { 
    Network, VerificationData, ProvingSystemId,
    getNextNonce, estimateFee, PriceEstimate,
    submitAndWaitVerification
} from "aligned-js";

const API_URL = 'http://localhost:3030';
const RPC_API = 'https://ethereum-holesky-rpc.publicnode.com';
const PRIVATE_KEY = '0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

async function openFile(elfFilePath) {
    const elfData = await fs.readFile(elfFilePath);
    return Array.from(elfData);
}

async function generateProof(elfFilePath, inputData) {
    const url = `${API_URL}/sp1/generate-proof-and-verify`;

    const elfData = await openFile(elfFilePath);
    const inputBytes = typeof inputData === 'string'
        ? Array.from(Buffer.from(inputData, 'utf-8'))
        : inputData;

    const data = { elf: elfData, input: inputBytes };

    const response = await axios.post(url, data);
    const responseJson = response.data;

    if (responseJson.message === 'Incorrect prove') {
        throw new Error('Incorrect prove!');
    } else if (responseJson.message === 'Output not added') {
        throw new Error('Output not added!');
    }

    return responseJson;
}

async function main() {
    const provider = new ethers.JsonRpcProvider(RPC_API);
    const wallet = new ethers.Wallet(PRIVATE_KEY, provider);

    const nonce = await getNextNonce(RPC_API, wallet.getAddress(), Network.Holesky);
    const maxFee = await estimateFee(RPC_API, PriceEstimate.Instant);
    const chainId = (await wallet.provider.getNetwork()).chainId;
    console.log(`Nonce: ${nonce}`);

    const root = path.join(__dirname, "..", "elf", "riscv32im-succinct-zkvm-elf");

    const proofResponse = await generateProof(root, 'cca');
    const proof = proofResponse.proof;
    const programCode = await openFile(root);

    const verificationData = VerificationData.new({
        provingSystem: ProvingSystemId.SP1,
        proof: proof,
        publicInput: null,
        verificationKey: null,
        vmProgramCode: programCode,
        proofGeneratorAddress: String(await wallet.getAddress())
    });

    const result = await submitAndWaitVerification(
        "wss://batcher.alignedlayer.com",
        RPC_API,
        Network.Holesky,
        verificationData,
        maxFee,
        wallet,
        nonce
    )
    console.log(`0x${Buffer.from(new Uint8Array(result.batchMerkleRoot)).toString('hex')}`);
}

main().catch((error) => {
    console.error('Error:', error);
});
